reflect.go

```diff
@@ -0,0 +1,387 @@
+				if fun != nil {
+					tf.checkCalls(fun)
+				}
+			}
+		}
+
+		if fun, ok := memb.(*ssa.Function); ok {
+			// these not only include top level functions, but also synthetic
+			// functions like the initialization of global variables
+			tf.checkCalls(fun)
+		}
+	}
+}
+
+func (tf *transformer) checkCalls(fun *ssa.Function) {
+	if fun.Synthetic == "loaded from gc object file" {
+		// fun.WriteTo crashes otherwise
+		return
+	}
+
+	/* fun.WriteTo(os.Stdout) */
+
+	var reflectParams []int
+
+	for _, block := range fun.Blocks {
+		for _, inst := range block.Instrs {
+			/* 	fmt.Printf("inst: %v, t: %T\n", inst, inst) */
+			call, ok := inst.(*ssa.Call)
+			if !ok {
+				continue
+			}
+			/* fmt.Printf("call: %v\n", call) */
+
+			callName := call.Call.Value.String()
+			if checkedAPIs[callName] {
+				// only check apis which were not already checked
+				continue
+			}
+
+			// record each call argument passed to a function parameter which is used in reflection
+			knownParams := cachedOutput.KnownReflectAPIs[callName]
+			for _, knownParam := range knownParams {
+				if len(call.Call.Args) <= knownParam {
+					continue
+				}
+
+				arg := call.Call.Args[knownParam]
+
+				/* fmt.Printf("flagging arg: %v\n", arg) */
+
+				visited := make(map[ssa.Value]bool)
+				reflectedParam := tf.recordArgReflected(arg, visited)
+				if reflectedParam == nil {
+					continue
+				}
+
+				pos := slices.Index(fun.Params, reflectedParam)
+				if pos < 0 {
+					continue
+				}
+
+				/* fmt.Printf("recorded param: %v func: %v\n", pos, fun) */
+
+				if !slices.Contains(reflectParams, pos) {
+					reflectParams = append(reflectParams, pos)
+				}
+			}
+		}
+	}
+
+	if len(reflectParams) > 0 {
+		cachedOutput.KnownReflectAPIs[fun.String()] = reflectParams
+
+		/* 	fmt.Printf("cachedOutput.KnownReflectAPIs: %v\n", cachedOutput.KnownReflectAPIs) */
+	}
+}
+
+// recordArgReflected finds the type(s) of a function argument, which is being used in reflection
+// and excludes these types from obfuscation
+// It also checks if this argument has any relation to a function paramter and returns it if found.
+func (tf *transformer) recordArgReflected(val ssa.Value, visited map[ssa.Value]bool) *ssa.Parameter {
+	// make sure we visit every val only once, otherwise there will be infinite recursion
+	if visited[val] {
+		return nil
+	}
+
+	/* fmt.Printf("val: %v %T %v\n", val, val, val.Type()) */
+	visited[val] = true
+
+	switch val := val.(type) {
+	case *ssa.IndexAddr:
+		for _, ref := range *val.Referrers() {
+			if store, ok := ref.(*ssa.Store); ok {
+				tf.recordArgReflected(store.Val, visited)
+			}
+		}
+		return tf.recordArgReflected(val.X, visited)
+	case *ssa.Slice:
+		return tf.recordArgReflected(val.X, visited)
+	case *ssa.MakeInterface:
+		return tf.recordArgReflected(val.X, visited)
+	case *ssa.UnOp:
+		return tf.recordArgReflected(val.X, visited)
+	case *ssa.FieldAddr:
+		return tf.recordArgReflected(val.X, visited)
```

Comment:
Maybe this can be replaced by `.Operands()`? It does exactly same thing and eliminates need for type switch

Reply:
I know this looks a bit repetitive but `.Operands` often doesn't do what I'm trying to achieve